core/chain_makers.go

```diff
@@ -372,13 +373,28 @@ func GenerateVerkleChain(config *params.ChainConfig, parent *types.Block, engine
 				kvs[string(key)] = v
 			}
 
+			// Initialize the preStateTrie if it is nil, this should
+			// correspond to the genesis block. This is a workaround
+			// needed until the main verkle PR is rebased on top of
+			// PBSS.
+			if preStateTrie == nil {
+				preStateTrie = vtr
+			}
+
 			vtr.Hash()
-			p, k, err := vtr.ProveAndSerialize(statedb.Witness().Keys(), kvs)
+			p, k, err := preStateTrie.ProveAndSerialize(statedb.Witness().Keys(), kvs)
 			if err != nil {
 				panic(err)
 			}
 			proofs = append(proofs, p)
 			keyvals = append(keyvals, k)
+
+			// save the current state of the trie for producing the proof for the next block,
+			// since reading it from disk is broken with the intermediate PBSS-like system we
+			// have: it will read the post-state as this is the only state present on disk.
+			// This is a workaround needed until the main verkle PR is rebased on top of PBSS.
```

Comment:
@gballet, sharing an idea....

Currently the `state_processor.go` pipeline is:
1. Execute the block.
2. Do migration of state (in case we're in the transition phase)
3. Finalize() (i.e: flush touched state in 1. to the tree)

We could add some step between 1. and 2. that loads the pre-state VKT since:
a. The tree on disk is still the pre-state tree (i.e: 3. hasn't run yet so is untouched)
b. The access-witness module already collected all the branches that were reads, so we have the keys.

So we can load the pre-state VKT, and at soon as is ready loaded from disk, then steps 2. and 3. can continue.
While 2. and 3. are running, we can generate the proof.

So this would allow us to do the proof generation overlapped with some other necessary steps (2. and 3.), maybe masking at least part (or all? don't know) of the proof building work.

Reply:
That makes sense, however the two following points need to be taken into consideration:

 * number 2. has to eventually be moved outside of `Process`  since we want to move the N values BEFORE the block is executed, so that only leaves 3. But if we look at it another way, it still holds: we do the proof generation more or less in line with 1. and 3., and it's 2. that we do in parallel, for the next block.
 * The question is whether the conversion data also needs to be in the block. My preference is that it does not, but it depends on what the sync needs.

Reply:
Cool, yeah we can re-think about this when we are sure about how the processing pipeline will work.